sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/NumberUtilsSuite.scala

```diff
@@ -0,0 +1,317 @@
+      "Format '9' used for parsing string to number or formatting number to string is invalid")
+    failParseWithInvalidInput(UTF8String.fromString("454"), "99",
+      "Format '99' used for parsing string to number or formatting number to string is invalid")
+
+    Seq(
+      ("454", "999") -> Decimal(454),
+      ("054", "999") -> Decimal(54),
+      ("404", "999") -> Decimal(404),
+      ("450", "999") -> Decimal(450),
+      ("454", "9999") -> Decimal(454),
+      ("054", "9999") -> Decimal(54),
+      ("404", "9999") -> Decimal(404),
+      ("450", "9999") -> Decimal(450)
+    ).foreach { case ((str, format), expected) =>
+      assert(parse(UTF8String.fromString(str), format) === expected)
+    }
+
+    failParseWithInvalidInput(UTF8String.fromString("454"), "0",
+      "Format '0' used for parsing string to number or formatting number to string is invalid")
+    failParseWithInvalidInput(UTF8String.fromString("454"), "00",
+      "Format '00' used for parsing string to number or formatting number to string is invalid")
+
+    Seq(
+      ("454", "000") -> Decimal(454),
+      ("054", "000") -> Decimal(54),
+      ("404", "000") -> Decimal(404),
+      ("450", "000") -> Decimal(450),
+      ("454", "0000") -> Decimal(454),
+      ("054", "0000") -> Decimal(54),
+      ("404", "0000") -> Decimal(404),
+      ("450", "0000") -> Decimal(450)
+    ).foreach { case ((str, format), expected) =>
+      assert(parse(UTF8String.fromString(str), format) === expected)
+    }
+
+    // Test '.' and 'D'
+    failParseWithInvalidInput(UTF8String.fromString("454.2"), "999",
+      "Format '999' used for parsing string to number or formatting number to string is invalid")
+    failParseWithInvalidInput(UTF8String.fromString("454.23"), "999.9",
+      "Format '999.9' used for parsing string to number or formatting number to string is invalid")
+
+    Seq(
+      ("454.2", "999.9") -> Decimal(454.2),
+      ("454.2", "000.0") -> Decimal(454.2),
+      ("454.2", "999D9") -> Decimal(454.2),
+      ("454.2", "000D0") -> Decimal(454.2),
+      ("454.23", "999.99") -> Decimal(454.23),
+      ("454.23", "000.00") -> Decimal(454.23),
+      ("454.23", "999D99") -> Decimal(454.23),
+      ("454.23", "000D00") -> Decimal(454.23),
+      ("454.0", "999.9") -> Decimal(454),
+      ("454.0", "000.0") -> Decimal(454),
+      ("454.0", "999D9") -> Decimal(454),
+      ("454.0", "000D0") -> Decimal(454),
+      ("454.00", "999.99") -> Decimal(454),
+      ("454.00", "000.00") -> Decimal(454),
+      ("454.00", "999D99") -> Decimal(454),
+      ("454.00", "000D00") -> Decimal(454),
+      (".4542", ".9999") -> Decimal(0.4542),
+      (".4542", ".0000") -> Decimal(0.4542),
+      (".4542", "D9999") -> Decimal(0.4542),
+      (".4542", "D0000") -> Decimal(0.4542),
+      ("4542.", "9999.") -> Decimal(4542),
+      ("4542.", "0000.") -> Decimal(4542),
+      ("4542.", "9999D") -> Decimal(4542),
+      ("4542.", "0000D") -> Decimal(4542),
```

Comment:
@beliefer I guess the issue is here: `,` at the end. Also fix the same below.

Reply:
Thank you for the guide.